### DISCUSSION TITLE
How to test state changes, e.g. screen rotations

### DISCUSSION THREAD

**Discussion by Burtan** [arkivanov/Decompose#333] [Q&A] [answered]
Hey,

as state changes are crucial events in all mobile apps, they should be tested thoroughly. I tried to emulate state changes by changing the lifecycle manually, e.g.:

```
fun LifecycleRegistry.emulateRotation() {
        pause()
        stop()
        destroy()
        create()
        start()
        resume()
    }
```

However, this does not work. I guess when destroy is called, in a real app the whole component tree is rebuild and with it a new context that is passed to your root component, controlling the lifecycle, state and instances. The way the component tree is rebuild is different per platform? The way the lifecycle is controlled, state and instances are persisted. This seems to be hard to emulate in a shared module.
I think I will have to use platform integration tests to test state changes? 
Maybe you could add such tests to your sample app, or you even succeed in emulating state changes in the shared module tests?

**Comment by arkivanov** [comment, marked as answer]
Hello! Do you mean testing component recreation? Basically, you should create a new instance of `DefaultComponentContext` and therefore a new instance of your component. There are some [sample tests](https://github.com/arkivanov/Decompose/blob/d708782e889771a5753be59134806d128acdcfe4/sample/shared/shared/src/commonTest/kotlin/com/arkivanov/sample/shared/counters/counter/CounterComponentTest.kt#L71).

**Comment by Burtan** [comment]
I didn't create an `InstanceKeeperDispatcher()` and a `StateKeeperDispatcher()` that is reused. Thanks for the clarification!
